api/extension-guides/language-model-tutorial.md

```diff
@@ -9,13 +9,325 @@ MetaDescription: Tutorial that walks you through creating a VS Code extension th
+
+We pass messages to the model in an array. This array can contain as many messages as you like. In our case, it contains the prompt followed by the users code with line numbers.
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getVisibleCodeWithLineNumbers(textEditor);
+
+  // select the 4o chat model
+  let [model] = await vscode.lm.selectChatModels({
+    vendor: 'copilot',
+    family: 'gpt-4o',
+  });
+
+  // init the chat message
+  const messages = [
+    vscode.LanguageModelChatMessage.User(ANNOTATION_PROMPT),
+    vscode.LanguageModelChatMessage.User(codeWithLineNumbers),
+  ];
+});
+```
+
+To send the messages to the model, we need to first make sure the selected model is available. This handles cases where the extension is not ready or the user is not signed in to GitHub Copilot. Then we send the messages to the model.
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getVisibleCodeWithLineNumbers(textEditor);
+
+  // select the 4o chat model
+  let [model] = await vscode.lm.selectChatModels({
+    vendor: 'copilot',
+    family: 'gpt-4o',
+  });
+
+  // init the chat message
+  const messages = [
+    vscode.LanguageModelChatMessage.User(ANNOTATION_PROMPT),
+    vscode.LanguageModelChatMessage.User(codeWithLineNumbers),
+  ];
+
+  // make sure the model is available
+  if (model) {
+
+    // send the messages array to the model and get the response
+    let chatResponse = await model.sendRequest(messages, {}, new vscode.CancellationTokenSource().token);
+
+    // handle chat response
+    await parseChatResponse(chatResponse, textEditor);
+  }
+});
+```
+
+Chat responses come in as fragments. These fragments usually contain single words, but sometimes they contain just punctuation. In order to display annotations as the response streams in, we want to wait until we have a complete annotation before we display it. Because of the way we have instructed our model to return its response, we know that when we see a closing `}` we have a complete annotation. We can then parse the annotation and display it in the editor.
+
+Add the missing `parseChatResponse` function above the `getVisibleCodeWithLineNumbers` method in the `extension.ts` file.
+
+```ts
+async function parseChatResponse(chatResponse: vscode.LanguageModelChatResponse, textEditor: vscode.TextEditor) {
+ let accumulatedResponse = "";
+
+ for await (const fragment of chatResponse.text) {
+  accumulatedResponse += fragment;
+
+  // if the fragment is a }, we can try to parse the whole line
+  if (fragment.includes("}")) {
+   try {
+    const annotation = JSON.parse(accumulatedResponse);
```

Comment:
this will be a good example case for JSON mode if we start supporting it
fyi @roblourens